feat: Add recurring scheduled flag changes

ordehi opened the pull request:
Adds ability to create recurring schedules for feature flag status changes. Schedules can repeat daily, weekly, or monthly.

- Model: Added is_recurring and recurrence_interval fields to ScheduledChange
- Task: Recurring schedules compute next run time instead of marking executed
- API: Validation requires interval when recurring, only update_status allowed
- Frontend: UI for checkbox + interval dropdown, table shows recurring indicator

<img width="1774" height="1184" alt="CleanShot 2025-12-11 at 19 29 06@2x" src="https://github.com/user-attachments/assets/bf267b2d-6628-4b85-98e2-811f97a5af06" />


- This covers use cases that can help workaround limitations like this: https://github.com/PostHog/posthog/issues/14796 (being able to update the flag on a repeating schedule mimics behavioral cohorts that check for rolling windows)
- And enables use cases like this user's: https://posthoghelp.zendesk.com/agent/tickets/44902

## How did you test this code?

Locally

## Changelog: (features only) Is this feature complete?

Yes

<!-- Yes if this is okay to go in the changelog. No if it's still hidden behind a feature flag, or part of a feature that's not complete yet, etc.  -->
<!-- Removing this section does not mean the changelog bot won't pick it up, because *some people* like to not use the template, so we can't rely on it existing. -->

ordehi committed:
feat: Add recurring scheduled flag changes

Adds ability to create recurring schedules for feature flag status changes.
Schedules can repeat daily, weekly, or monthly.

- Model: Added is_recurring and recurrence_interval fields to ScheduledChange
- Task: Recurring schedules compute next run time instead of marking executed
- API: Validation requires interval when recurring, only update_status allowed
- Frontend: UI for checkbox + interval dropdown, table shows recurring indicator

Note: Migration needs to be generated (python manage.py makemigrations posthog)

Tests: Added 5 task tests + 6 API tests

ordehi committed:
make migration

ordehi committed:
test snapshot

ordehi committed:
move repeat checkbox next to date dropdown, give better recurring interval description

ordehi committed:
Merge origin/master into feat/recurring-scheduled-flag-changes

dmarticus commented on posthog/tasks/process_scheduled_changes.py:
```diff
@@ -59,6 +61,17 @@ def is_unrecoverable_error(exception: Exception) -> bool:
+        return current + relativedelta(weeks=1)
+    elif interval == "monthly":
+        return current + relativedelta(months=1)
+    raise ValueError(f"Unknown recurrence interval: {interval}")
```
Using `relativedelta(months=1)` handles most cases, but consider documenting or testing behavior for edge cases like scheduling on Jan 31st (would go to Feb 28/29).

dmarticus reviewed (changes requested):
This is a great change! In addition to the greptile comments (I think they're all valid to address), there's a few more things I'd like this change to do before it's ready to release:

- For recurring schedules, it might be useful to track when the schedule last ran successfully (separate from the next `scheduled_at`). Consider tracking a `last_executed_at`; this would help with debugging and audit trails.
- Consider adding an optional `end_date` or `max_occurrences` field for users who want time-bounded recurrence.
- And the biggest one: there's no way to stop/delete recurring schedules from UI. The existing delete functionality should work, but there's no explicit "pause" or "stop recurrence" option. Users might want to temporarily disable a recurring schedule without deleting it.

ordehi committed:
feat(scheduled-changes): Add recurring schedule model fields

- Add is_recurring boolean field
- Add recurrence_interval choice field (daily/weekly/monthly)
- Add last_executed_at for tracking successful recurring executions

ordehi committed:
fix(scheduled-changes): Backend bug fixes for recurring schedules

- Fix delayed execution: skip ahead if next_run is still in past
- Update last_executed_at on successful recurring execution
- Validate recurrence_interval is a valid choice (daily/weekly/monthly)
- Fix validation to work on both create and update operations
- Add last_executed_at to API response (read-only)
- Add comprehensive docstring for compute_next_run

ordehi committed:
feat(scheduled-changes): UI improvements for recurring schedules

- Fix monthly display: show "last day" for days 29-31 to avoid confusion
- Improve checkbox label: "Make recurring" instead of "Repeat this change"
- Add "Stop recurring" option to More menu for active recurring schedules
- Add updateScheduledChange API method for PATCH operations

ordehi committed:
feat(scheduled-changes): UI improvements for recurring schedules

- Fix monthly display: show "last day" for days 29-31 to avoid confusion
- Improve checkbox label: "Make recurring" instead of "Repeat this change"
- Add "Pause recurring" / "Resume recurring" options to More menu
- Paused schedules (is_recurring=false + recurrence_interval) don't execute
- Preserve recurrence_interval when pausing (allows resume)
- Add status badges: "Recurring" for active, "Paused" for paused schedules
- Add updateScheduledChange API method for PATCH operations

ordehi committed:
feat(scheduled-changes): Add optional end_date for time-bounded recurrence

- Add end_date field to ScheduledChange model
- Stop recurring when next_run exceeds end_date (marks as completed)
- Add end_date picker in UI (optional, shown when recurring is enabled)
- Show end_date in table tooltip for active recurring schedules

ordehi committed:
fix: Use correct Dayjs type import for kea-typegen

ordehi committed:
Merge branch 'master' into feat/recurring-scheduled-flag-changes

dmarticus commented on frontend/src/scenes/feature-flags/featureFlagLogic.ts:
```diff
@@ -677,6 +683,35 @@ export const featureFlagLogic = kea<featureFlagLogicType>([
+            {
+                setRecurrenceInterval: (_, { interval }) => interval,
+                // Reset when recurring is disabled or operation changes
+                setIsRecurring: (state, { isRecurring }) => (isRecurring ? state : null),
```
when `isRecurring` is set to false, `recurrenceInterval` is reset to null, but this conflicts with how we "pause" execution in the backend where `recurrence_interval` is preserved when `is_recurring` is false. If a user creates a recurring schedule, pauses it, then tries to create a new non-recurring schedule, the old interval might leak through. The frontend should distinguish between "creating new" vs "pausing existing."

dmarticus commented on posthog/tasks/process_scheduled_changes.py:
```diff
@@ -80,9 +117,32 @@ def process_scheduled_changes() -> None:
+                        # to avoid immediate re-trigger or missed executions piling up
+                        now = timezone.now()
+                        while next_run <= now:
+                            next_run = compute_next_run(next_run, scheduled_change.recurrence_interval)
```
Is this the behavior we want? Even if it is, it's worth logging when we're skipping executions and maybe adding a metric on this count so we can observe it with prometheus/grafana.

e.g. something like this
```
    # If task execution was delayed and next_run is still in the past, skip ahead
    now = timezone.now()
    skipped_count = 0
    while next_run <= now:
        next_run = compute_next_run(next_run, scheduled_change.recurrence_interval)
        skipped_count += 1
    
    if skipped_count > 1:
        # We expected to skip 1 (the one we just executed), but if we skipped more,
        # that means executions were missed due to worker delay
        missed_count = skipped_count - 1
        logger.warning(
            f"Recurring schedule {scheduled_change.id} missed {missed_count} "
            f"execution(s) due to delayed processing. "
            f"Interval: {scheduled_change.recurrence_interval}, "
            f"Original scheduled_at: {scheduled_change.scheduled_at}, "
            f"Next run: {next_run}"
        )
```

we could _also_ add a model field in the DB like this if we wanted

```
# In the model
missed_execution_count = models.IntegerField(default=0)

# In the task
if missed_count > 0:
    scheduled_change.missed_execution_count = F('missed_execution_count') + missed_count
    logger.warning(...)
```

probably too much overhead though.

ordehi replied:
Nice yeah, added tracking for this.

dmarticus commented on posthog/tasks/test/test_process_scheduled_changes.py:
```diff
@@ -1107,3 +1107,161 @@ def test_cannot_schedule_change_in_past(self) -> None:
+            feature_flag.refresh_from_db()
+            # Schedule corrects the manual override
+            self.assertEqual(feature_flag.active, False)
+            self.assertEqual(scheduled_change.scheduled_at, datetime(2024, 1, 17, 9, 0, tzinfo=UTC))
```
I think you're missing one test here: tests around paused behavior. Maybe add something like this?

```py
# Add to test_process_scheduled_changes.py
@freeze_time("2024-01-15T09:00:00Z")
def test_paused_recurring_schedule_does_not_execute(self) -> None:
    """Paused recurring schedules (is_recurring=False with interval) should be skipped."""
    feature_flag = FeatureFlag.objects.create(
        name="Paused Flag",
        key="paused-flag",
        active=True,
        filters={"groups": []},
        team=self.team,
        created_by=self.user,
    )

    scheduled_change = ScheduledChange.objects.create(
        team=self.team,
        record_id=feature_flag.id,
        model_name="FeatureFlag",
        payload={"operation": "update_status", "value": False},
        scheduled_at=datetime(2024, 1, 15, 9, 0, tzinfo=UTC),
        is_recurring=False,  # Paused
        recurrence_interval="daily",  # But has interval
    )

    process_scheduled_changes()

    scheduled_change.refresh_from_db()
    feature_flag.refresh_from_db()
    # Should NOT have executed
    self.assertIsNone(scheduled_change.executed_at)
    self.assertEqual(feature_flag.active, True)  # Unchanged
```

dmarticus reviewed (approved):
For what it's worth; I think this is actually good to approve – I'd just appreciate if you considered my comments first.

Overall this is sick though; thank you!

Commit:
Merge remote-tracking branch 'origin/master' into feat/recurring-scheduled-flag-changes

# Conflicts:
#	posthog/migrations/max_migration.txt

ordehi committed:
fix: Renumber migration to 0944 after merge conflict with master

The recurring scheduled changes migration was originally 0940, but master
now has 0940-0943 for other features. Renumbered to 0944 to resolve conflict.

ordehi committed:
refactor: Add ScheduledChangeOperationType enum and use in API validation

Replaces magic string "update_status" with enum constant for better
type safety and maintainability.

ordehi committed:
fix: Add validation that end_date must be after scheduled_at

ordehi committed:
docs: Clarify interval validation location in compute_next_run docstring

ordehi committed:
feat: Add logging and Prometheus metric for missed scheduled executions

When a recurring schedule's execution is delayed and we need to skip ahead
multiple intervals, we now log a warning and increment a Prometheus counter
for observability.

ordehi committed:
docs: Add comment explaining why one-time schedules don't set last_executed_at

ordehi committed:
fix: Use end-of-day for end_date to avoid timezone issues

When a user selects an end date with day granularity, we now use
endOf('day') to ensure the schedule runs through the entire selected
day regardless of the user's timezone.

ordehi committed:
refactor: Remove implicit reducer side effects, use explicit resets

- Remove setIsRecurring side effects from recurrenceInterval and endDate reducers
- Add explicit setEndDate(null) to createScheduledChangeSuccess handler
- Makes reset behavior explicit and removes confusing implicit coupling

ordehi committed:
test: Add test for paused recurring schedule behavior

Verifies that schedules with is_recurring=False but recurrence_interval
preserved are treated as "paused" and do not execute.

ordehi commented on frontend/src/scenes/feature-flags/featureFlagLogic.ts:
```diff
@@ -1103,6 +1138,9 @@ export const featureFlagLogic = kea<featureFlagLogicType>([
                         scheduled_at: scheduleDateMarker.toISOString(),
+                        is_recurring: values.isRecurring,
+                        recurrence_interval: values.recurrenceInterval,
+                        end_date: values.endDate?.toISOString() || null,
```
Oh noes! Went with end-of-day, makes the most sense to me.

ordehi commented on posthog/tasks/process_scheduled_changes.py:
```diff
@@ -59,6 +61,35 @@ def is_unrecoverable_error(exception: Exception) -> bool:
     return isinstance(exception, unrecoverable_types)
 
 
+def compute_next_run(current: datetime, interval: str) -> datetime:
```
We don't seem to use `Literal` but good idea to type this, so I'll break the ice.

ordehi committed:
feat(scheduled-changes): Use Literal type for recurrence interval parameter

ordehi committed:
fix(scheduled-changes): Reject creating non-recurring schedules with interval set

ordehi committed:
fix(scheduled-changes): UI improvements for recurring schedules

- Use project timezone for end date instead of local timezone
- Add tooltip explaining end-of-day semantics with link to settings
- Add End date column to scheduled changes table
- Always show interval dropdown (disabled when not recurring) to prevent layout shift
- Clear interval and end date when unchecking Make recurring

ordehi committed:
Merge branch 'master' into feat/recurring-scheduled-flag-changes

ordehi committed:
Merge branch 'master' into feat/recurring-scheduled-flag-changes

Resolve migration conflict by renaming 0944 to 0945

ordehi committed:
fix(scheduled-changes): Revert to str type for interval parameter

Django's TextChoices fields return str at runtime, not the enum type,
causing mypy errors when using Literal. Added docstring explaining this
limitation - validation is done at the API layer via the serializer.
